src/chain/shared_state.rs

```diff
@@ -292,3 +321,66 @@ impl Debug for SectionProofChain {
         )
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::{chain::SectionInfo, BlsPublicKey, FullId, Prefix, XorName};
+    use std::collections::BTreeSet;
+    use unwrap::unwrap;
+
+    fn gen_section_info(pfx: Prefix<XorName>) -> SectionInfo {
+        let sec_size = 5;
+        let mut members = BTreeSet::new();
+        for _ in 0..sec_size {
+            let id = FullId::within_range(&pfx.range_inclusive());
+            let _ = members.insert(*id.public_id());
+        }
+        unwrap!(SectionInfo::new(members, pfx, None))
+    }
+
+    fn gen_pk(pfx: Prefix<XorName>) -> BlsPublicKey {
+        BlsPublicKey::from_section_info(&gen_section_info(pfx))
+    }
+
+    #[test]
+    fn test_update_their_keys() {
+        let empty: Prefix<XorName> = Default::default();
+        let pfx_0 = empty.pushed(false);
+        let pfx_1 = empty.pushed(true);
+        let pfx_10 = pfx_1.pushed(false);
+        let pfx_11 = pfx_1.pushed(true);
+        let pfx_100 = pfx_10.pushed(false);
+        let pfx_101 = pfx_10.pushed(true);
+
+        // our section is 0
+        let sec_0 = gen_section_info(pfx_0);
+        let mut state = SharedState::new(sec_0);
+
+        // add keys for neighbours 10 and 11
+        let pk_sec_10 = gen_pk(pfx_10);
+        let pk_sec_11 = gen_pk(pfx_11);
+        state.update_their_keys(pfx_10, pk_sec_10.clone());
+        state.update_their_keys(pfx_11, pk_sec_11.clone());
+
+        assert_eq!(state.get_their_key(&pfx_10), Some(&pk_sec_10));
+        assert_eq!(state.get_their_key(&pfx_11), Some(&pk_sec_11));
+
+        // add key for neighbour 101 - this means that 10 has split, so we should no longer have a
+        // key for 10, but we should assign the old 10 key to 100 now
+        let pk_sec_101 = gen_pk(pfx_101);
+        state.update_their_keys(pfx_101, pk_sec_101.clone());
```

Comment:
I would probably have tried to organize it so I can use an AAA pattern.
Maybe something that would look a bit like:

```
fn update_their_keys(update_prefix: Vec<Prefix>, expected_prefix:Vec<(Prefix, Prefix)) {
    let prefix_to_keys: BTreeMap<_,_> = update_prefix.map(|prefix| (*prefix, gen_pk(prefix)).collect();
    let sec_0 = gen_section_info(pfx_0);
    let mut state = SharedState::new(sec_0);
    
    for prefix in update_prefix {
          state.update_their_keys(prefix, unwrap!(prefix_to_keys.get(prefix)));
    }

    let expected: BTreeMap<_,_> = expected_prefix.map(|pfx, pfx_key| (*pfx, unwrap!(prefix_to_keys.get(pfx_key))).collect();
   assert!((state.get_their_key(), expected);
}

#[test]
fn test_update_their_keys_one() {
    update_their_keys([empty, pfx_10], [(pfx_10, pfx_10), (pfx_0, pfx_0), (pfx_11, empty)]);
}

fn test_update_their_keys_thee_bit_one_after_the_other() {
  ....
}
...
````